app: Update Forall command to allow multiple concurrent processes

pdgendt opened the pull request:
Demonstrate asynchronous behavior for the Forall command and add an argument to select the number of jobs.

The same idea can be applied to the `Update` command.

pdgendt committed:
app: Update Forall command to allow multiple concurrent processes

Demonstrate asynchronous behavior for the Forall command and add an
argument to select the number of jobs.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt commented:
For comparison, timing counting the lines of code using [`cloc`](https://github.com/AlDanial/cloc) with 20 logical cores:

```
west forall -c "cloc --quiet ."  103.21s user 5.00s system 70% cpu 2:34.15 total
west forall -j -c "cloc --quiet ."  148.83s user 5.32s system 472% cpu 32.600 total
```

pdgendt committed:
tests: Add test for 'forall' with jobs

Add test cases for running the forall command with multiple processes.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

marc-hb reviewed (commented):
This looks really cool, thanks for designing and testing this! I'm afraid we still have the issue of concurrent outputs though, see below.

marc-hb commented on src/west/app/project.py:
```diff
@@ -1670,16 +1671,15 @@ def do_add_parser(self, parser_adder):
         parser.add_argument('projects', metavar='PROJECT', nargs='*',
                             help='''projects (by name or path) to operate on;
                             defaults to active cloned projects''')
+        parser.add_argument('-j', '--jobs', nargs='?', const=-1,
```
```suggestion
        # Default to 1 when `-j` is not given because there is no way to
        # whether the user commands can be run at the same time safely.
        parser.add_argument('-j', '--jobs', nargs='?', const=-1,
```
Eventually, `west grep`, `west update` and others could default to `cpu_count()` if everything goes well but I think `forall` should always default to 1.

(such a comment also helps a bit with the peculiar `default+const` argparse idiom)

marc-hb reviewed (dismissed):
Important process output questions to address.

pdgendt commented on src/west/app/project.py:
```diff
@@ -1689,12 +1689,24 @@ def do_run(self, args, user_args):
+            self.banner(f'running "{args.subcommand}" in {project.name_and_path}:')
+            proc = await asyncio.create_subprocess_shell(args.subcommand,
+                                                         cwd=cwd, env=env, shell=True)
+            return await proc.wait()
```
Yes, I didn't want to dive right in here, but it would be nice to get right from the get-go.

I was dabbling with the idea of having the output behave like `ninja` where the output line is replaced with the banner and maybe some counter indicating the progress. And when the subprocesses is done, print its output as is.

Not sure if this would require something like `curses`, I think it could be more lightweight.

marc-hb replied:
> I was dabbling with the idea of having the output behave like ninja where the output line is replaced with the banner and maybe some counter indicating the progress. And when the subprocesses is done, print its output as is

That would be awesome but I think just 1) making sure the output is readable 2) all commands use the same output "framework" would already be a major milestone and great stepping stone towards something better. And it would give what a lot of users have been waiting for: concurrency at last.

pdgendt committed:
app: Add optional end argument to banner print methods

Allow passing a custom end to banner methods. This is useful for example
to print a carriage return.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

marc-hb reviewed (commented):
> You probably tested this with relatively "quiet" and easy to read output and a reasonable number of threads... can you try again after cranking it all up? This code must be prepared to handle not just the "common" cases but all cases.

OK, I just had a test idea and this PR was pretty easy to "break" after all. The following one-liner prints readable output with `-j 1` and totally jumbled up with `-j > 1`

```shell
west forall -j 1 -c 'i=8; while test $((i--)) -ge 0;
 do printf " $WEST_PROJECT_NAME"; sleep 0.1; done; printf "\n"'
```

It would be great to "upgrade" a test like this and make it part to the actual test suite. I don't know how we could make it  portable to Windows.. by converting it to Python maybe? Or maybe we don't need to? I think it would be OK to have _some_ tests skipped on some platforms, there would still be value in that.

Don't get me wrong: we DO need some tests that exert the terminal on Windows. Windows terminal issues was the reason for the 7223431da114 revert and that whole saga. But maybe not _all_ tests need to run on all platforms.

pdgendt commented:
The updated version is better, as all printing is done from a single thread, I tried to overwrite the "running" line, but it doesn't clear it.

marc-hb commented:
Colorization was surprisingly more robust than I thought (mainly due to a lack of "toggles") but I managed to craft a test that broke your earlier version. But you just fixed it before I shared the test :-) I'm sharing that test code anyway because I think it's still useful:

```
west forall -j 10 -c 'i=7; while test $((--i)) -ge 0;
 do printf "\e[$((30+i))m"; sleep 0.0${RANDOM}; printf " $WEST_PROJECT_NAME-$i"; done; printf "\e[0m\n"'
```

<img width="1022" alt="Screenshot 2024-10-18 at 10 14 13" src="https://github.com/user-attachments/assets/b58eee8a-55eb-436c-a971-51e8dfdde0f0">

pdgendt commented:
> Colorization was surprisingly more robust than I thought (mainly due to a lack of "toggles") but I managed to craft a test that broke your earlier version. But you just fixed it before I shared the test :-) I'm sharing that test code anyway because I think it's still useful:
> 
> ```
> west forall -j 10 -c 'i=7; while test $((--i)) -ge 0;
>  do printf "\e[$((30+i))m"; sleep 0.0${RANDOM}; printf " $WEST_PROJECT_NAME-$i"; done; printf "\e[0m\n"'
> ```

Which shell do you use? Doesn't work properly with `zsh`.

marc-hb commented:
> Which shell do you use? Doesn't work properly with zsh.

The default/standard: bash. Just type `bash` to switch temporarily. 

Should be compatible with all Bourne shells: ash, ksh, etc.

Strange `zsh` went its own way...

marc-hb commented:
> Which shell do you use? Doesn't work properly with zsh.

In `zsh`, try: `emulate ksh`, this should get close enough to POSIX.
https://zsh.sourceforge.io/Doc/Release/Invocation.html#Compatibility
https://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html

What does not work?

I just checked and `$(( ))` is POSIX https://mywiki.wooledge.org/ArithmeticExpression
On the other hand, `${RANDOM}` is most likely a bashism: replace it with 2 or 3.

A must have for any shell script: https://www.shellcheck.net/ (`apt install shellcheck`)

marc-hb commented:
> "A shell" is vague. On Linux this is apparently the default shell. Why not but this must be more explicit. Can you please update this help text? 

So the reason this is "vague" is because 1) it's a mess 2) accordingly, the Python documentation is vague too - intentionally! https://github.com/python/cpython/issues/114539

So I think we only need a warning that defers to Python documentation here. But we do need that warning. Something like "it's a wildly non-portable and insecure mess, check the Python documentation and don't use this in automation". Something like that.

pdgendt commented:
Very outdated, and probably not something I'll pick back up soon.
